### ISSUE TITLE
vanity gen for TON blockchain

### ISSUE THREAD

**Issue by boo50** [10gic/vanitygen-plusplus#1]
this is not an issue actually, but inquiry to add TON blockchain to vanity gen
https://github.com/ton-blockchain/ton

**Comment by 10gic** [comment]
Telegram has shut down its TON cryptocurrency project. 😭

Currently, vanitygen plusplus can support a new blockchain very easily, if the chain:
1. use elliptic curve secp256k1, and
2. its address encoded with Base58Check.

It seems that TON does not meet the conditions, https://ton.live/accounts?section=all
To support it, may be there is extra work to be done.

**Comment by boo50** [comment]
Yes, but there are Free TON project which is maintained by https://github.com/tonlabs/ and it's live, even mainnet already active. And ton.live is actually their website.

**Comment by 10gic** [comment]
Is there technical article about how to generating TON address? Just like https://en.bitcoin.it/wiki/Technical_background_of_version_1_Bitcoin_addresses

**Comment by boo50** [comment]
Yes sure on the top of howto
https://ton.org/HOWTO.txt

**Comment by 10gic** [comment]
> Yes sure on the top of howto
> https://ton.org/HOWTO.txt

It just talks about how to convert address from raw form (eg. -1:fcb91a3a3816d0f7b8c2c76108b8a9bc5a6b7a55bd79f8ab101c52db29232260) to user-friendly forms.
There is no guide about how to generate an address inside masterchain (workchain_id=-1).

**Comment by boo50** [comment]
Yes you right.
Maybe this pdf starting from page 53 have necessary information?
https://test.ton.org/tblkch.pdf

**Comment by 10gic** [comment]
@boo50 I'm not familiar with TON.

In source code of tonos-cli https://github.com/tonlabs/tonos-cli/blob/master/src/genaddr.rs, the key part of generating address is:
```
    let addr = ton.contracts.get_deploy_address(
        abi.clone().into(),
        &contract,
        initial_data.clone().map(|d| d.into()),
        &keys.public,
        wc,
    ).map_err(|e| format!("failed to generate address: {}", e.to_string()))?;
```
Is the address always related with contract?

**Comment by boo50** [comment]
Yes, every address in ton is actually a smart-contract.
This one for example https://github.com/tonlabs/ton-labs-contracts/tree/master/solidity/safemultisig

**Comment by 10gic** [comment]
@boo50  Does it mean that user must provide a concrete contract before generating address?
Given a private/public key-pair and a contract, is the address determined?

**Comment by boo50** [comment]
yes, generate address looks like this
https://github.com/tonlabs/net.ton.dev/blob/master/scripts/msig_genaddr.sh
in this example address is generating for contract provided before - safemultisig

**Comment by 10gic** [comment]
Done in https://github.com/10gic/vanitygen-plusplus/commit/81592c1c131965e83937673aed4030b4b84912a5
